### ISSUE TITLE
Files in a/ directory miss their directory prefix

### ISSUE THREAD

**Issue by lode** [so-fancy/diff-so-fancy#205]
I've got a change in a file called `a/index.php`.

With `git --no-pager diff` I get:

```
diff --git a/index.php a/index.php
```

dsf shows:

```
modified: index.php
```

**Comment by stevemao** [comment]
Because you specify `--no-pager` so the diff is not piped to `diff-so-fancy`

**Comment by lode** [comment]
I don't understand.

I have a directory called `a`. Within that directory is a file called `index.php`. When I change that file, I expect to see a diff naming the `a/index.php` path. However, dsf skips the `a/` directory. So the output of `--no-pager` is what I expect. Dsf should show `modified: a/index.php`.

**Comment by scottchiefbaker** [comment]
Git prepends "fake" directories on each diff. Specifically it puts `a/dir/filename` and `b/dir/filename` for the changed files. We [strip that off](https://github.com/so-fancy/diff-so-fancy/blob/master/libexec/diff-so-fancy.pl#L71) because it's not a real directory. It would make sense that you're seeing the `a/` get stripped off... but I can't recreate what you're seeing.

I recreated your scenario as follows:

1. mkdir a/
2. cat "bar" > a/foo
3. git commit -a
4. cat "barz" > a/foo

Git diff raw shows (note the double a's in "a/a/foo"):

```
:git --no-pager diff
diff --git a/a/foo b/a/foo
index 5716ca5..ae7bf01 100644
--- a/a/foo
+++ b/a/foo
@@ -1 +1 @@
-bar
+barz
```

When run through diff-so-fancy I get:

```
─────────────
modified: a/foo
─────────────
@ foo:1 @
-bar
+barz
```

In other words... it works for me. Are you using the latest version from the master git branch?

**Comment by lode** [comment]
I found the issue, I have `[diff] noprefix = true` in my gitconfig. Don't know where I got that one from, but it explains the behavior of the normal git diff.

```
$ git --no-pager diff
diff --git a/foo a/foo
index 5716ca5..78509d0 100644
--- a/foo
+++ a/foo
@@ -1 +1 @@
-bar
+baf
$ git diff
─────────────────────────────────────────
modified: foo
─────────────────────────────────────────
@ foo:1 @
-bar
+baf
```

I would be cool of course if dfs could recognize that setting.

**Comment by stevemao** [comment]
@lode Makes sense. PR welcome :)

**Comment by scottchiefbaker** [comment]
Just so I understand... if you see `noprefix = true` diff won't append `a/` and `b/` at the beginning of each changed file? 

If that's the case (and you want to keep this setting) it's probably easiest to modify your d-s-f alias to disable that feature at the CLI, while leaving it in your config?

Otherwise... d-s-f would have to be able to detect if this is enabled and act appropriately. 

This is pretty corner case:

1. You have to have `noprefix = true`
2. Your changed file also has to be in a `a/` or `b/` dir

**Comment by OJFord** [comment]
So essentially `noprefix = true` gives the behaviour of removing `a/`, `b/` without dsf.

I wonder what the value of git config's Boolean options are if unset by the user - if they're actually ternary, it might be nice to essentially change the default behaviour (i.e. `noprefix` unset) so that we do (pseudocode):

```
if config.diff.noprefix in {true, unset}:
    removeAB()
```

> This is pretty corner case

A related issue would be "I have `diff.noprefix = false` in my config, but `a/` and `b/` are missing from the output". The scheme above would also nip that in the bud.

**Comment by lode** [comment]
I just found that I created thus issue myself by requesting these prefixes to be removed: https://github.com/so-fancy/diff-so-fancy/issues/28#issuecomment-182293866

> <sub>I have [diff] noprefix = true in my gitconfig. Don't know where I got that one from</sub>

<sub>Now I do :)</sub>

And that at that moment it was already noted these two would conflict. But also that the PR removing the prefixes *shouldn't* conflict with the git setting. Apparently it does?

**Comment by scottchiefbaker** [comment]
Where did we leave this? Is this something we should try and fix, or is this a "feature".

**Comment by OJFord** [comment]
My view is that `diff.noprefix` should be respected, for the case I mentioned above where it's explicitly set `false`. DSF should act as configured, but just use a different default to git; strip prefix if `noprefix` unset.

**Comment by scottchiefbaker** [comment]
Ok we'll have to check for that setting in `~/.gitconfig` and act accordingly. I think I can do that.

Adding a test might be difficult though.

**Comment by OJFord** [comment]
> Ok we'll have to check for that setting in ~/.gitconfig and act accordingly.

That's not the guaranteed location (e.g. `$XDG_DATA_DIRECTORY/git/config` is another possibility, and whatever Windows does with `%APPDATA%` I think) - it's best I think to do `git config diff.noprefix` to get the output as `[nothing]`/`false`/`true`.

I haven't looked at the tests in a while, but with the above it ought to then be possible to mock (i.e. temporarily replace I suppose) `git-config`.

**Comment by scottchiefbaker** [comment]
Good point. We already have `git_config_boolean()` which handles reading whatever config git uses, so I'll use that for this.

https://github.com/so-fancy/diff-so-fancy/blob/master/libexec/diff-so-fancy.pl#L243

**Comment by scottchiefbaker** [comment]
@OJFord / @stevemao if one of you can test this and let me know if it works for you (all tests pass), then I will land this code.
